File: cole/day43_evaluate-reverse-polish-notation/evaluate-reverse-polish-notation.py
# the slow but easy python solution
from typing import List

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        i = 2
        while len(tokens) > 1:
            if tokens[i] in '+-*/':
                expr = f'({tokens[i-2]}){tokens[i]}({tokens[i-1]})'
                tokens = tokens[:i-2] + [int(eval(expr))] + \
                    (tokens[i+1:] if i < len(tokens) - 1 else [])
                i -= 2
            i += 1
        return int(tokens[0])

File: cole/day43_evaluate-reverse-polish-notation/evaluate-reverse-polish-notation.c
// the "fast" solution with a stack in C

int evalRPN(char ** tokens, int tokensSize) {
    int stack[tokensSize], ptr = 0;
    int x;
    for (int i = 0; i < tokensSize; ++i) {
        if (sscanf(tokens[i], "%d", &x)) { stack[ptr++] = x; continue; }
        switch (*tokens[i]) {
            case '+':
                stack[ptr - 2] += stack[ptr - 1];
                break;
            case '-':
                stack[ptr - 2] -= stack[ptr - 1];
                break;
            case '*':
                stack[ptr - 2] *= stack[ptr - 1];
                break;
            case '/':
                stack[ptr - 2] /= stack[ptr - 1];
                break;
            default:
                perror("what in tarnation");
                exit(1);
        }
        --ptr;
    }
    return stack[0];
}